atomicapp/nulecule/main.py

```diff
@@ -0,0 +1,234 @@
+        self.nulecule = None
+
+    def unpack(self, update=False):
+        if not os.path.isdir(self.unpack_path) or update:
+            self.nulecule = Nulecule.unpack(self.image, self.unpack_path)
+
+    def install(self):
+        self.unpack()
+        self.nulecule.install()
+
+    def run(self, provider_key=None, dry=False):
+        self.install()
+        self.nulecule.run(provider_key, dry)
+
+    def stop(self, provider_key=None):
+        self.nulecule.stop(provider_key)
+
+    def uninstall(self):
+        self.stop()
+        self.nulecule.uninstall()
+
+    def clean(self, force=False):
+        self.uninstall()
+        shutil.rmtree(self.unpack_path)
+        self.initialize()
+
+
+class Nulecule(NuleculeBase):
+    """
+    This represents an application compliant with Nulecule specification.
+    A Nulecule instance can have instances of Nulecule and Nulecule as
+    components. A Nulecule instance knows everything about itself and its
+    componenents, but does not have access to it's parent's scope.
+    """
+    def __init__(self, id, specversion, metadata, graph, basepath,
+                 requirements=None, params=None, config=None,
+                 namespace=GLOBAL_CONF):
+        self.id = id
+        self.specversion = specversion
+        self.metadata = metadata
+        self.graph = graph
+        self.basepath = basepath
+        self.requirements = requirements
+        self.params = params or []
+        self.namespace = namespace
+        self.config = None
+
+    @classmethod
+    def unpack(cls, image, path, config={}, namespace=GLOBAL_CONF):
+        docker_handler = DockerHandler()
+        docker_handler.pull(image)
+        docker_handler.extract(image, APP_ENT_PATH, path)
+        nulecule_data = anymarkup.parse_file(
+            os.path.join(path, 'Nulecule'))
+        nulecule = Nulecule(config=config, basepath=path, namespace=namespace,
+                            **nulecule_data)
+        nulecule.load_components()
+        return nulecule
+
+    def install(self):
+        self.load_config()
+        self.render()
+
+    def run(self, provider_key=None, dry=False):
+        provider_key, provider = self.get_provider(provider_key, dry)
+        for component in self.components:
+            component.run(provider_key, dry)
+
+    def stop(self, provider_key=None):
+        provider_key, provider = self.get_provider(provider_key)
+        # stop the Nulecule application
+        for component in self.components:
+            component.stop(provider_key)
+
+    def uninstall(self):
+        # uninstall the Nulecule application
+        for component in self.components:
+            component.uninstall()
+
+    def load_config(self, config={}):
+        super(Nulecule, self).load_config(config=config)
+        for component in self.components:
+            _config = {}
+            _config[GLOBAL_CONF] = copy.deepcopy(
+                self.config.get(GLOBAL_CONF)) or {}
```

Comment:
bug here `copy.deepcopy(self.config.get(GLOBAL_CONF)) or {}` is not what you want. `copy.deepcopy(self.config.get(GLOBAL_CONF) or {}))` is what you want I think